### PR TITLE
pkp/pkp-lib#6685 Add tombstonemanager for OPS

### PR DESCRIPTION
**Pull request by ajnyga**
https://github.com/pkp/pkp-lib/issues/6685

### CODE REVIEW

File: classes/submission/Repository.inc.php [hunk 1]
```diff
@@ -13,8 +13,36 @@
...
+            $tombstoneDao->deleteByDataObjectId($submission->getId());
+        } elseif ($oldStatus === Submission::STATUS_PUBLISHED && $newStatus !== $oldStatus) {
+            $requestContext = $this->request->getContext();
+            if ($requestContext && $requestContext->getId() === $submission->getData('contextId')) {
+                $context = $requestContext;
+            } else {
+                $context = Services::get('context')->get($submission->getData('contextId'));
+            }
+            import('classes.preprint.PreprintTombstoneManager');
+            $preprintTombstoneManager = new \PreprintTombstoneManager();
```
Comment (asmecher) [inline classes/submission/Repository.inc.php:44]: It should be possible to use a fully qualified classname (`use APP\preprint\PreprintTombstoneManager;`) and then remove the `import` call.

**Review by asmecher — CHANGES_REQUESTED** [review]
Thanks, @ajnyga! I made some comments on the PR.

File: classes/preprint/PreprintTombstoneManager.inc.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+class PreprintTombstoneManager
+{
+    /**
+     * Constructor
+     */
+    public function __construct()
+    {
+    }
+
+    public function insertPreprintTombstone(&$preprint, &$context)
```
[the remark this answers is not in the document]
Comment (ajnyga) [reply classes/preprint/PreprintTombstoneManager.inc.php:35]: are these not needed for the hook there?

File: classes/preprint/PreprintTombstoneManager.inc.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        $sectionDao = DAORegistry::getDAO('SectionDAO'); /* @var $sectionDao SectionDAO */
+        $tombstoneDao = DAORegistry::getDAO('DataObjectTombstoneDAO'); /* @var $tombstoneDao DataObjectTombstoneDAO */
+        // delete preprint tombstone -- to ensure that there aren't more than one tombstone for this preprint
+
+        $tombstoneDao->deleteByDataObjectId($preprint->getId());
+        // insert preprint tombstone
+        $section = $sectionDao->getById($preprint->getSectionId());
+        $setSpec = urlencode($context->getPath()) . ':' . urlencode($section->getLocalizedAbbrev());
+        $oaiIdentifier = 'oai:' . Config::getVar('oai', 'repository_id') . ':' . 'preprint/' . $preprint->getId();
+        $OAISetObjectsIds = [
```
[the remark this answers is not in the document]
Comment (ajnyga) [reply classes/preprint/PreprintTombstoneManager.inc.php:46]: check

File: classes/preprint/PreprintTombstoneManager.inc.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                ->filterByContextIds([$contextId])
+                ->filterByStatus([Submission::STATUS_PUBLISHED])
+        );
+        foreach ($submissions as $submission) {
+            $tombstoneDao->deleteByDataObjectId($submission->getId());
+        }
+    }
+}
+
+if (!PKP_STRICT_MODE) {
```
[the remark this answers is not in the document]
Comment (ajnyga) [reply classes/preprint/PreprintTombstoneManager.inc.php:99]: check

File: classes/submission/Repository.inc.php (hunk 1 above)
Comment (ajnyga) [reply classes/submission/Repository.inc.php:44]: check

File: classes/preprint/PreprintTombstoneManager.inc.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajnyga) [reply classes/preprint/PreprintTombstoneManager.inc.php:35]: I think the only plugin I have seen using the hook was the old openaire plugin. I will remove the references, but note that OJS still has those in the same code.

[the remark this answers is not in the document]
Comment (asmecher) [reply classes/preprint/PreprintTombstoneManager.inc.php]: Thanks! I've removed the references from the OJS code.

**Comment by asmecher** [comment]
Merged -- thanks, @ajnyga!

**Pull request merged**